.github/workflows/compile.yml

```diff
@@ -2,44 +2,141 @@ name: CI Emscripten
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - id: set-matrix
+        run: echo "::set-output name=matrix::$(bash ./build.sh --list --core-names-only | jq -Rsc '. / "\n" - [""]')"
 
-      - name: Generate checksums
-        uses: jmgilman/actions-generate-checksum@v1
+  build-core:
+    name: Build Core
+    needs:
+      - build-environment
+      - get-jobs
+    runs-on: [self-hosted, linux, x64 ]
+    continue-on-error: true
+    strategy:
+      matrix:
+        core: ${{fromJson(needs.get-jobs.outputs.matrix)}}
+    steps:
+      - name: Build Cores
+        run: |
+          source "emsdk/emsdk_env.sh"
+          bash build.sh --core=${{ matrix.core }}
+      - name: Move core
+        run: |
+          mkdir -p ./artifacts
+          mv ./output/${{ matrix.core }}.zip ./artifacts/${{ matrix.core }}.zip
+      - uses: actions/attest-build-provenance@v2
         with:
-          patterns: |
-            ./compile/EmulatorJS/data/cores/*.data
+          subject-path: './artifacts/${{ matrix.core }}.zip'
+  
+  package:
+    name: Package
+    needs: build-core
+    runs-on: [self-hosted, linux, x64 ]
+    steps:
+      - name: Zip files
+        run: |
+          cd ./output
+          zip -r '../artifacts/cores.zip' * -x '*.zip'
+          cd ..
+      - uses: actions/attest-build-provenance@v2
+        with:
+          subject-path: './artifacts/cores.zip'
 
-      - name: Get Version
-        id: version
-        run: echo "VERSION=$(cat VERSION)" >> "$GITHUB_OUTPUT"
+  release:
+    name: Release
+    needs: package
+    runs-on: [self-hosted, linux, x64 ]
+    if: startsWith(github.ref, 'refs/tags/')
+    steps:
+      - name: Upload Release Asset
+        uses: softprops/action-gh-release@v2
+        with:
+          files: artifacts/cores.zip
 
-      - name: Create Release & Make Notes & Upload Assets
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          VERSION: ${{ steps.version.outputs.VERSION }}
+  upload-nightly-cdn:
+    name: Upload nightly to CDN
+    needs:
+      - get-jobs
+      - build-core
+    runs-on: [self-hosted, linux, x64 ]
+    steps:
+      - name: Setup SSH
+        run: |
+          echo "${{secrets.SSH_PRIVATE_KEY}}" > ~/.ssh/id_rsa
+          echo "Host ${{ secrets.SSH_HOST }}" > ~/.ssh/config
+          echo "  User ${{ secrets.SSH_USERNAME }}" >> ~/.ssh/config
+          echo "  IdentityFile ~/.ssh/id_rsa" >> ~/.ssh/config
+          chmod 600 ~/.ssh/id_rsa
+      - name: Replace remote files
         run: |
-          gh release create "v$VERSION" --title "Latest Version v$VERSION" --generate-notes
-          gh release upload "v$VERSION" ./compile/EmulatorJS/data/cores/*.data
-          gh release upload "v$VERSION" "EmulatorJS Cores.zip"
-          gh release upload "v$VERSION" "checksum.txt#checksum-256.txt"
+          rsync -Pav -e "ssh -i $HOME/.ssh/id_rsa" ./output/ ${{ secrets.SSH_USERNAME}}@${{secrets.SSH_HOST}}:/
```

Comment:
Maybe add `z` to the flags (enables compression). No need to waste bandwidth on a 200mb text file that can be compressed to 30mb